// @intellectronica/ruler@0.3.35/dist/mcp/propagateOpenCodeMcp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.propagateMcpToOpenCode = propagateMcpToOpenCode;
const fs = __importStar(require("fs/promises"));
const FileSystemUtils_1 = require("../core/FileSystemUtils");
const path = __importStar(require("path"));
function isLocalServer(value) {
    const server = value;
    return (server &&
        (typeof server.command === 'string' || Array.isArray(server.command)));
}
function isRemoteServer(value) {
    const server = value;
    return server && typeof server.url === 'string';
}
/**
 * Transform ruler MCP configuration to OpenCode's specific format
 */
function transformToOpenCodeFormat(rulerMcp) {
    const rulerServers = rulerMcp.mcpServers || {};
    const openCodeServers = {};
    for (const [name, serverDef] of Object.entries(rulerServers)) {
        const openCodeServer = {
            type: 'local',
            enabled: true,
        };
        if (isRemoteServer(serverDef)) {
            openCodeServer.type = 'remote';
            openCodeServer.url = serverDef.url;
            if (serverDef.headers) {
                openCodeServer.headers = serverDef.headers;
            }
            if (typeof serverDef.timeout === 'number') {
                openCodeServer.timeout = serverDef.timeout;
            }
        }
        else if (isLocalServer(serverDef)) {
            openCodeServer.type = 'local';
            const command = Array.isArray(serverDef.command)
                ? serverDef.command
                : [serverDef.command];
            const args = serverDef.args || [];
            openCodeServer.command = [...command, ...args];
            if (serverDef.env) {
                openCodeServer.environment = serverDef.env;
            }
            if (typeof serverDef.timeout === 'number') {
                openCodeServer.timeout = serverDef.timeout;
            }
        }
        else {
            continue;
        }
        openCodeServers[name] = openCodeServer;
    }
    return {
        $schema: 'https://opencode.ai/config.json',
        mcp: openCodeServers,
    };
}
async function propagateMcpToOpenCode(rulerMcpData, openCodeConfigPath, backup = true) {
    const rulerMcp = rulerMcpData || {};
    // Read existing OpenCode config if it exists
    let existingConfig = {};
    try {
        const existingContent = await fs.readFile(openCodeConfigPath, 'utf8');
        existingConfig = JSON.parse(existingContent);
    }
    catch {
        // File doesn't exist, we'll create it
    }
    // Transform ruler MCP to OpenCode format
    const transformedConfig = transformToOpenCodeFormat(rulerMcp);
    // Merge with existing config, preserving non-MCP settings
    const finalConfig = {
        ...existingConfig,
        $schema: transformedConfig.$schema,
        mcp: {
            ...existingConfig.mcp,
            ...transformedConfig.mcp,
        },
    };
    await (0, FileSystemUtils_1.ensureDirExists)(path.dirname(openCodeConfigPath));
    if (backup) {
        const { backupFile } = await Promise.resolve().then(() => __importStar(require('../core/FileSystemUtils')));
        await backupFile(openCodeConfigPath);
    }
    await fs.writeFile(openCodeConfigPath, JSON.stringify(finalConfig, null, 2) + '\n');
}
